internal/proxy/util.go

```diff
@@ -944,6 +945,49 @@ func checkPrimaryFieldData(schema *schemapb.CollectionSchema, insertMsg *msgstre
 	return ids, nil
 }
 
+// TODO(smellthemoon): can merge it with checkPrimaryFieldData
+func upsertCheckPrimaryFieldData(schema *schemapb.CollectionSchema, result *milvuspb.MutationResult, insertMsg *msgstream.InsertMsg) (*schemapb.IDs, error) {
+	rowNums := uint32(insertMsg.NRows())
+	// TODO(dragondriver): in fact, NumRows is not trustable, we should check all input fields
+	if insertMsg.NRows() <= 0 {
+		return nil, errNumRowsLessThanOrEqualToZero(rowNums)
+	}
+
+	if err := checkLengthOfFieldsData(schema, insertMsg); err != nil {
+		return nil, err
+	}
+
+	primaryFieldSchema, err := typeutil.GetPrimaryFieldSchema(schema)
+	if err != nil {
+		log.Error("get primary field schema failed", zap.String("collectionName", insertMsg.CollectionName), zap.Any("schema", schema), zap.Error(err))
+		return nil, err
+	}
+
+	// get primaryFieldData whether autoID is true or not
+	var primaryFieldData *schemapb.FieldData
+	if primaryFieldSchema.AutoID {
+		// upsert has not supported when autoID == true
```

Comment:
xinguo 你好，我在这里有些小疑问，请问下为何要禁止 auto id 为 True 呢? 
如果判断为 upsert，复用原来的 primary keys （RowIDs），代码逻辑上似乎也是可以支持的。
@smellthemoon

Reply:
Because upsert is actually set to insert first and then delete (the logic here is that although delete and insert are in the same msgPack, when applying the messages in msgPack, they are still done separately, and the atomicity of this step in querynode is not guaranteed. If do insert first, it is equivalent to writing the data to the growing segment first, and now there is a guarantee that when searching, this data must be searched), based on this logic, you can probably understand why it is set to not allow autoid==true , if autoid==true, pk cannot be obtained through fielddata in the delete phase.

Reply:
@KubrickLiu

Reply:
Even if set auto id == true, can pk be reused in insert sub-process of the upsert process.

Reply:
Acutally, Im not sure completely where to set pk , is in proxy or datanode , could you tell me
@smellthemoon

Reply:
If specified that Upsert request should fill pk field although set auto id == true, could we solved 【pk cannot be obtained through fielddata in the delete phase.】

Reply:
> If specified that Upsert request should fill pk field although set auto id == true, could we solved 【pk cannot be obtained through fielddata in the delete phase.】

The pk were set by users, if autoid==true, pk will set by autoGenPrimaryFieldData() in insert. I am not sure what you mean[specified that Upsert request should fill pk field although set auto id == true], if set autoid==true but you still input pk field,  upsert or insert both will error. The pk unchanged all the time in upsert(one of the cores to keep upsert atomicity), when checkPrimaryFieldData in upsert  is used to prevent errors.

Reply:
> Even if set auto id == true, can pk be reused in insert sub-process of the upsert process.
> @smellthemoon

I am not sure what it means. Pk will keep unchanged all the time in upsert. There is a one-to-one correspondence between pk and field data. You can simply think that if you don't specify pk, then the bottom layer will not know what data you want to update. If you specify a pk, but this pk does not have corresponding data in the original system, then it is considered that you want to insert these data.

Reply:
> xinguo 你好，我在这里有些小疑问，请问下为何要禁止 auto id 为 True 呢?
> 如果判断为 upsert，复用原来的 primary keys （RowIDs），代码逻辑上似乎也是可以支持的。
> @smellthemoon

The problem is, we can We can find data by pk, but not pk by data.

Reply:
抱歉我可能没有很理解您的意思，【if autoid==true, pk cannot be obtained through fielddata in the delete phase】关于这里的 auto id 为 true，为何 delete 阶段会拿不到 pk 呢？即使 auto id 为true，是不是也可以保持 upsert 阶段的 insert 过程保持原 pk 不变呢？
抱歉我可以和您微信沟通十分钟吗，感谢 ~